content/client-lib-development-guide/features.textile

```diff
@@ -488,6 +490,12 @@ h3(#realtime-channel). Channel
 ** @(RTL4f)@ Once an @ATTACH@ @ProtocolMessage@ is sent, if an @ATTACHED@ @ProtocolMessage@ is not received within the "default realtime request timeout":#defaults, the attach request should be treated as though it has failed and the channel should transition to the @SUSPENDED@ state. The channel will then be subsequently automatically re-attached as described in "RTL13":#RTL13
 ** @(RTL4d)@ If the language permits, a callback can be provided that is called when the channel is attached successfully or the attach fails and the @ErrorInfo@ error is passed as an argument to the callback
 ** @(RTL4e)@ If the user does not have sufficient permissions to attach to the channel, the channel will transition to @FAILED@ and set the @Channel#errorReason@
+** @(RTL4j)@ If the attach is not a clean attach (where a 'clean attach' is an attach attempt where the channel has either not previously been attached or has been explicitly detached since the last time it was attached), for example an automatic reattach triggered by "@RTN15c3@":#RTN15c3 or "@RTL13a@":#RTL13a (non-exhaustive), the library should set the "@ATTACH_RESUME@":#TR3f flag in the @ATTACH@ message
```

Comment:
Can we not state more explicitly, in terms of channel state, when the flag is set?

Reply:
@paddybyers Not really, as unfortunately it's not just a function of the current and previous channel state. Consider the following two sequences of states:

1. initialized, `attach()` called -> attaching -> suspended -> attaching -> attached

2. attached, server-initiated detach received -> attaching -> suspended -> attaching -> attached

in both cases the last ATTACH message is sent when going from `suspended` to `attaching`, but in the first case ATTACH_RESUME should be set, and in the second case it should not be.

So I'm not sure there is a nicer way to express the condition than `a 'clean attach' is an attach attempt where the channel has either not previously been attached or has been explicitly detached since the last time it was attached`, you do need a separate boolean to keep track of it. (Unless we add eg `attaching_unclean` and `suspended_unclean` states, which is horrible)

Reply:
I could be more explicit about needing a boolean variable and what exact conditions should set it to what though, if you think that'd be better?

Reply:
@paddybyers https://github.com/ably/docs/pull/766/commits/d6a3cbd5a0c9cc334a11c47ac3dab52ad5bb5c11